### ISSUE TITLE
V16 base_url adaptation

### ISSUE THREAD

**Issue by sebastienbeau** [shopinvader/odoo-shopinvader#1373]
## Context 

**base_url** will need some change as before the record that have the url was the "shopinvader.product" or the "shopinvader.category" and on this object the field lang_id was defined and required.

As we drop this model, some adaptation need to be done.
Before doing the change, it's time to ask us some question.

Also the current url implementation on product and category is done in the shopinvader module. It's should be done in some extra module (see https://github.com/shopinvader/odoo-shopinvader/issues/1371) 

## Question 1:

Url must be uniq for a website, what should be our “referential” on V16 ?
- the sale channel
- the index
- global

Pro-against

If it’s the sale channel, this mean that in case you want to share an index between two sale channel we are blocked as we can have a different url per sale channel and only one value in the index

If it’s the index, this mean that you depend the search-engine stack to have url on your product / category (maybe not an issue)

If it’s global: this means that you share the same url in different websites and you also share the old redirection between all your sites (when you publish an old product on a new website it will have redirect url). This have no impact but just add some extra info in the index.

## Question 2:

When should we generate the url key ?

### Immediately (current implementation)

For example in the case of the product.
When you change the name of the product a new url key is immediately generated (with redirection), if you rename it again a new url will be created even if the data even been yet synchronize (so we can have useless redirection).
Same with category (but category have parent and url depend on parent, so changing parent update url on all children)

### When needed
An alternative will to just flag the current url as obsolete and when we generate the json for the search engine we generate the new url (with the redirection)

## My feedback is the following

Question 1:
By default it’s should be global and if you need an extra module can generate url per index (this module can be done later)

Question 2:
It’s should be generated at json recomputation so it avoid useless url generation

## Module naming

I propose

shopinvader_base_url
shopinvader_product_url
shopinvader_category_url
shopinvader_url_per_index

Does we should prefix this module with shopinvader ? or this module are enough generic and should be move in oca in server-tools / product-attribute / search-engine

Do you have a better name ?

What is your point of view ? Maybe I miss something ?

@sbidoul @lmignon @hparfr

**Comment by sebastienbeau** [comment]
@simahawk

**Comment by lmignon** [comment]
@sebastienbeau Thank you for the deep dive on this matter.  I agree partially with your feedback . 
Others strategies should be implemented in dedicated modules. Nevertheless, computing the value at json recomputation introduce an implicit dependency on an other addon. We could make the url field a computed one and have a mdel serializer helper module providing a method to add the url into the json. At the same time, this method could maintain the  redirection table and the current url. If the current url is != than the url from the model it will update the redirection table etc....
To summarize, 

base_url,
model_url_serializer_helper
shopinvader_product_serializer (where product_product extend url.mixin and we implement the serializer for the required attributes used by the shopinvader frontend)
shopinvader_category_serializer (where product_category extend url.mixin and we implement the serializer for the required attributes used by the shopinvader frontend)

:thinking: :thinking: :thinking: 
....

**Comment by sebastienbeau** [comment]
I miss to give some extra info.

For the generation at the last time, I will keep the logic of computed field.

So we will have two field on my record (product.template and product.category)
- url_key
- redirect_url_key : currently defined [here](https://github.com/shopinvader/odoo-shopinvader/blob/f0656b7e579f2771a13503f7eca66fd6a5de879a/shopinvader/models/shopinvader_binding.py#LL18C5-L18C21)

Both are computed field note stored (and not visible in the view) that depend on the context (by default depend on the lang but can depend on the index with the additional module)

Then the seralizer will just read that field, so we are not dependent of the serializer implementation.

**Comment by lmignon** [comment]
These fields are only for our UCs. We need these since the navigation depends on this url into the index. IMO they should only be implemented into the serializers...

**Comment by sbidoul** [comment]
> Does we should prefix this module with shopinvader ? or this module are enough generic and should be move in oca in server-tools / product-attribute / search-engine

I see value in prefixing all addons in the shopinvader repos with `shopinvader_`, and rename them if/when they migrate to OCA. So it's easy to know where they come from, and avoid potential name clashes.

For the rest I don't know the subject well enough to have an opinion :)
